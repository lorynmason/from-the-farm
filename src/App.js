import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import { Buy } from './Components/Buy';

class App extends Component {
  render() {
    return (
      <div className="App">
        <Buy />
      </div>
    );
  }
}

export default App;
