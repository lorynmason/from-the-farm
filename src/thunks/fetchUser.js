import { isLoading, hasErrored, addUser } from '../actions';
import { cleanUser } from '../helpers/cleaner'

export const fetchUser = (url, token) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch(url);
      if (!response.ok) {
        throw Error(response.statusText);
      }
      dispatch(isLoading(false))
      const results = await response.json();
      const user = cleanUser(results.data, token);
      dispatch(addUser(user));
    } catch (error) {
      console.log(error)
      dispatch(hasErrored(error.message));
    }
  }
}