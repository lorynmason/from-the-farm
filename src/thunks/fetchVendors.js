import { isLoading, hasErrored, addVendors, addProducts } from '../actions';

export const fetchVendors = () => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch('https://xpoll-be.herokuapp.com/api/v1/vendors');
      if (!response.ok) {
        throw Error(response.statusText);
      }
      dispatch(isLoading(false))
      const results = await response.json();
      const vendors = cleanVendors(results.data);
      const products = cleanProducts(results.data);
      dispatch(addVendors(vendors));
      dispatch(addProducts(products));
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}

export const cleanVendors = (vendors) => {
  return vendors.map( vendor => {
    const { name, address, state, city, phone, email, lat, long, bio } = vendor.attributes
    return {
      name,
      bio,
      address,
      state, 
      city,
      phone,
      email,
      lat,
      long, 
      id: parseInt(vendor.id)
    }
  })
}
  
export const cleanProducts = (vendors) => {
  return vendors.reduce((productArr, vendor) => {
    const { name, products } = vendor.attributes;
    products.reduce((obj, product) => {
      const itemName = Object.keys(product)[0];

      obj = {
        vendorName: name,
        name: Object.keys(product)[0],
        ...product[itemName]
      }
      productArr.push(obj);
      return obj
    }, {})

    return productArr;
  }, [])
}
