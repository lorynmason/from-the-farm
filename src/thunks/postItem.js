import { isLoading, addMessage, hasErrored } from '../actions';

export const postItem = (url, item, userToken) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          "Content-Type": "application/json",
          "Accept": "application/json",
          "Authorization": `Bearer ${userToken}`
        },
        body: JSON.stringify(item)
      });
      if (!response.ok) {
        throw Error(response.statusText);
      }
      dispatch(isLoading(false));
      const results = await response.json();
      const newItem = results.data.attributes;
      dispatch(addMessage(`Successfully added ${newItem.name} at ${newItem.price} per ${newItem.unit}`));
    } catch(error) {
      dispatch(hasErrored(error.message));
      dispatch(addMessage('Unable to add product'));
    }
  }
}