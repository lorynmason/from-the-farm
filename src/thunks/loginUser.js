import { isLoading, hasErrored } from '../actions';
import { fetchUser } from './fetchUser'

export const loginUser = (url, state) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          email: state.email,
          password: state.password
        })
      });
      if (!response.ok) {
        throw Error(response.statusText);
      }
      dispatch(isLoading(false))
      const results = await response.json();
      dispatch(fetchUser(`https://xpoll-be.herokuapp.com/api/v1/users/${results.user.id}`, results.auth_token)) 
    } catch (error) {
      dispatch(hasErrored(error.message));
    }
  }
}