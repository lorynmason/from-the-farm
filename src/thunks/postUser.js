import { isLoading, hasErrored } from '../actions';
import { fetchUser } from './fetchUser'

export const postUser = (url, state) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true));
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          email: state.email,
          password: state.password,
          name: state.name,
          password_confirmation: state.password_confirmation,
          account_type: state.account_type,  
          address: state.address,
          city: state.city,
          state: state.state,
          phone: state.phone,
          zip: state.zip,
          img_url: state.img_url,
          bio: state.bio
        })
      });
      if (!response.ok) {
        throw Error(response.statusText);
      }
      dispatch(isLoading(false))
      const results = await response.json();
      dispatch(fetchUser(`https://xpoll-be.herokuapp.com/api/v1/users/${results.user.id}`, results.auth_token)) 
    } catch (error) {
      dispatch(hasErrored(error.message));
    }
  }
}