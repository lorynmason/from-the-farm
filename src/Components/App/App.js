import React, { Component } from 'react';
import { Buy } from '../Buy/Buy';
import { Header } from '../Header/Header';
import '../../styles/main.scss';
import { Switch, Route, withRouter } from 'react-router';
import Profile from '../Profile/Profile';
import { createBrowserHistory } from 'history';
import { connect } from 'react-redux';
import { addUser, removeUser, addProducts, addMessage } from '../../actions';
import { fetchVendors } from '../../thunks/fetchVendors';
import Message from '../Message/Message';

class App extends Component {

  async componentDidMount() {
    this.props.fetchVendors('https://xpoll-be.herokuapp.com/api/v1/vendors');
    this.props.addUserToStore({
      address: "85 Hooker St",
      bio: "Orange You Glad You Didn't Say Bananas",
      city: "denver",
      email: "ruffnbuff@example.com",
      id: 6,
      lat: 39.717646,
      long: -105.029438,
      name: "Oranges 4 Eva",
      phone: 2313414141,
      state: "CO"
    })
  }

  search = async({ productId, location, range }) => {
    const baseUrl = 'https://xpoll-be.herokuapp.com/api/v1/search'
    let path;
    if (!location) {
      console.log('error: please enter a search location')
      this.props.addMessage('error: please enter a search location')
    } 
    
    if (!range) {
      range = "50"
    } 
    
    if (!productId) {
      path = `?loc=${location}&range=${range}`
    } else {
      path = `?loc=${location}&range=${range}&item=${productId}`
    }
  
    const response = await this.props.fetchVendors(baseUrl + path)
    this.filterProducts(productId, response)
  }

  filterProducts = (id, response) => {
    if (id) {
      const newProducts = this.props.products.filter( product => {
        id = parseInt(id)
        return product.item_id === id
      });
      
      this.props.addProductsToStore(newProducts);
    }
  }

  render() {
    const history = createBrowserHistory();
    return (
      <div className="App">
        <Header />
        <Message />
        <Switch>
          <Route path="/buy" render={() => <Buy history={history} search={this.search}/>}/>
          <Route path="/profile" render={() => <Profile user={this.props.user} products={this.props.products} />}/>
        </Switch>
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  vendors: state.vendors,
  products: state.products,
  user: state.user
});

export const mapDispatchToProps = (dispatch) => {
  return {
    addUserToStore: (user) => dispatch(addUser(user)),
    fetchVendors: (url) => dispatch(fetchVendors(url)),
    addProductsToStore: (products) => dispatch(addProducts(products)),
    addMessage: (message) => dispatch(addMessage(message))
  } 
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));
