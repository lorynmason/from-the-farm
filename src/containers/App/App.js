import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Switch, Route, withRouter, Redirect } from 'react-router';
import Buy from '../Buy/Buy';
import AddProductForm from '../AddProductForm/AddProductForm';
import Header from '../Header/Header';
import '../../styles/main.scss';
import Profile from '../Profile/Profile';
import { fetchVendors } from '../../thunks/fetchVendors';
import Message from '../Message/Message';
import { About } from '../About/About';
import Login from '../../containers/Login/Login';

export class App extends Component {
  async componentDidMount() {
    this.props.fetchVendors('https://xpoll-be.herokuapp.com/api/v1/vendors');
  }

  render() {
    return (
      <div className="App">
        <Header />
        <Message />
        <Switch>
          <Route exact path="/" component={Buy} />
          <Route path="/about" component={About}/>
          <Route path="/login" component={Login}/>
          <Route path="/buy" component={Buy} />
          <Route path="/add-product" render={() => this.props.user.name ? (<AddProductForm />) : (<Redirect to="/login" />)} />
          <Route path="/profile" render={() => this.props.user.name ? (<Profile />) : (<Redirect to="/login" />)} />
        </Switch>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  vendors: state.vendors,
  products: state.products,
  user: state.user,
  isLoading: state.isLoading
});

export const mapDispatchToProps = (dispatch) => {
  return {
    fetchVendors: url => dispatch(fetchVendors(url))
  };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));
