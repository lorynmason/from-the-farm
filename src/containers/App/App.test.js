import React from 'react';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { shallow } from 'enzyme';
import { fetchVendors } from '../../thunks/fetchVendors';

jest.mock('../../thunks/fetchVendors.js');

describe('App', () => {
  let wrapper;
  let mockDispatch;
  beforeEach(() => {
    wrapper = shallow(<App fetchVendors={fetchVendors}/>);
    mockDispatch = jest.fn();
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  it('should call fetchVendors with the correct params', async () => {
    await wrapper.instance().componentDidMount();

    expect(fetchVendors).toHaveBeenCalledWith('https://xpoll-be.herokuapp.com/api/v1/vendors');
  });

  describe('mapStateToProps', () => {
    it('should return a props object', () => {
      const mockGlobalState = {
        vendors: [],
        products: [],
        user: {},
        isLoading: false,
        vendorSearchResults: [],
        productSearchResults: []
      }

      const expected = {
        vendors: [],
        products: [],
        user: {},
        isLoading: false
      }

      const mappedProps = mapStateToProps(mockGlobalState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should dispatch fetchVendors thunk with the correct args when fetchVendors is called', () => {
      const mockUrl = 'someplace.com';
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchVendors(mockUrl);
      expect(mockDispatch).toHaveBeenCalledWith(fetchVendors(mockUrl));
    }); 
  });
});
