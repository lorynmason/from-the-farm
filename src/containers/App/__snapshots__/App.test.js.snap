// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    fetchVendors={
      [MockFunction] {
        "calls": Array [
          Array [
            "https://xpoll-be.herokuapp.com/api/v1/vendors",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(Header) />,
        <Connect(Message) />,
        <Switch>
          <Route
            component={[Function]}
            exact={true}
            path="/"
          />
          <Route
            component={[Function]}
            path="/about"
          />
          <Route
            component={[Function]}
            path="/login"
          />
          <Route
            component={[Function]}
            path="/buy"
          />
          <Route
            path="/add-product"
            render={[Function]}
          />
          <Route
            path="/myprofile"
            render={[Function]}
          />
          <Route
            path="/profile/:id"
            render={[Function]}
          />
        </Switch>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Route
              component={[Function]}
              exact={true}
              path="/"
            />,
            <Route
              component={[Function]}
              path="/about"
            />,
            <Route
              component={[Function]}
              path="/login"
            />,
            <Route
              component={[Function]}
              path="/buy"
            />,
            <Route
              path="/add-product"
              render={[Function]}
            />,
            <Route
              path="/myprofile"
              render={[Function]}
            />,
            <Route
              path="/profile/:id"
              render={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "path": "/about",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "path": "/login",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "path": "/buy",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "path": "/add-product",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "path": "/myprofile",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "path": "/profile/:id",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(Header) />,
          <Connect(Message) />,
          <Switch>
            <Route
              component={[Function]}
              exact={true}
              path="/"
            />
            <Route
              component={[Function]}
              path="/about"
            />
            <Route
              component={[Function]}
              path="/login"
            />
            <Route
              component={[Function]}
              path="/buy"
            />
            <Route
              path="/add-product"
              render={[Function]}
            />
            <Route
              path="/myprofile"
              render={[Function]}
            />
            <Route
              path="/profile/:id"
              render={[Function]}
            />
          </Switch>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Route
                component={[Function]}
                exact={true}
                path="/"
              />,
              <Route
                component={[Function]}
                path="/about"
              />,
              <Route
                component={[Function]}
                path="/login"
              />,
              <Route
                component={[Function]}
                path="/buy"
              />,
              <Route
                path="/add-product"
                render={[Function]}
              />,
              <Route
                path="/myprofile"
                render={[Function]}
              />,
              <Route
                path="/profile/:id"
                render={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "exact": true,
                "path": "/",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "path": "/about",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "path": "/login",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "path": "/buy",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "path": "/add-product",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "path": "/myprofile",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "path": "/profile/:id",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
